The ::class constant (#104)
WordPress/wpcs-docs: wordpress-coding-standards/php.md

[/warning]

While rare, `array` and class/interface name based type declarations already exist in WordPress Core.

The function signature of any function (method) which can be overloaded by plugins or themes should not be touched.
This includes all existing `public` and `protected` class methods and any conditionally declared (pluggable) functions in the global namespace.

This leaves, for now, only unconditionally declared functions in the global namespace, `private` class methods, and code newly being introduced as candidates for adding type declarations.

For those, adding `callable`, class and interface name based parameter type declarations where the parameter type only expects one type and it would allow for simplification of existing/new code, is allowed and tentatively encouraged.

Using the `array` keyword in type declarations is **strongly discouraged** for now, as most often, it would be better to use `iterable` to allow for more flexibility in the implementation and that keyword is not yet available for use in WordPress Core until the minimum requirements are raised to PHP 7.1.

### The ::class constant

> When using the `::class` constant for class name resolution you should add no space between the `::class` and the preceding class name it applies to. There should be no space between the double colon and the `class` keyword, and the `class` keyword should be in lowercase.

Reviewer: ```suggestion
When using the `::class` constant for class name resolution, the `class` keyword should be in lowercase and there should be no spaces around the `::` operator.
```

```php
// Correct.
add_action( 'action_name', array( My_Class::class, 'method_name' ) );
 
// Incorrect.
add_action( 'action_name', array( My_Class :: CLASS, 'method_name' ) );
```

_Usage in WordPress Core_

The `::class` constant can be used in WordPress Core.

> Replacing existing uses of `__CLASS__` with `self::class` and `get_called_class()` with `static::class`, where appropriate, is encouraged.

Reviewer: As there are no PHP cross-version issues/considerations, I don't think this needs to be mentioned in the coding standards docs.

Maybe we should just open up an issue in Trac for fixing these instead ? (if there isn't one open already) 

FYI: I'm actually coaching some people on writing a sniff to detect and auto-fix these. May still be a while before the sniff is available (in PHPCSExtra), but once it is, I'd suggest adding it to the `WordPress-Extra` ruleset.

If we wait a little with opening the Trac ticket, we could just auto-generate a patch instead of waiting for people to manually create patches ;-)

Author: We can then just remove the entire Usage in WordPress Core part? I don't think mentioning that it can be used in the core has any merit - we wouldn't really mention a rule in the handbook if it cannot be used in the core in some form, no?

Reviewer: I agree with removing the "Usage in Core" part. We do mention some features in the CS docs which cannot be yet used in Core and for those having a section about "Usage in Core" makes sense. For things where there are no PHP cross-version considerations, I don't think we need to say that explicitly.

(In a way, that would be like saying: you can use arithmetic operators 😂 )
